File: util/src/main/java/jcog/pri/PriReference.java
package jcog.pri;

import jcog.Util;
import jcog.math.FloatSupplier;

import java.util.Arrays;
import java.util.function.BiConsumer;
import java.util.function.Supplier;

/**
 * prioritized reference
 */
public interface PriReference<X> extends Prioritizable, Supplier<X>, FloatSupplier {

    static float[] histogram(Iterable<? extends Prioritized> pp, float[] x) {
        var bins = x.length;
        double[] total = {0};

        for (Prioritized y : pp) {
            if (y == null)
                continue;
            var p = y.priElseZero();
            if (p > 1f) p = 1f;
            var b = Util.bin(p, bins);
            x[b]++;
            total[0]++;
        }

        var t = total[0];
        if (t > 0) {
            for (var i = 0; i < bins; i++)
                x[i] /= t;
        }
        return x;
    }

    /**
     * double[histogramID][bin]
     */
    static <X, Y> double[][] histogram( Iterable<PriReference<Y>> pp,  BiConsumer<PriReference<Y>, double[][]> each,  double[][] d) {

        for (var y : pp) {
            each.accept(y, d);
        }

        for (var e : d) {
            var total = Arrays.stream(e).sum();
            if (total > 0) {
                for (int i = 0, eLength = e.length; i < eLength; i++) {
                    var f = e[i];
                    e[i] /= total;
                }
            }
        }

        return d;
    }

    @Override
    default float asFloat() {
        return pri();
    }
}

File: util/src/main/java/jcog/pri/PriReference.kt
package jcog.pri

import jcog.Util
import jcog.math.FloatSupplier

import java.util.function.BiConsumer
import java.util.function.Supplier

/**
 * prioritized reference
 */
interface PriReference<X> : Prioritizable, Supplier<X>, FloatSupplier {

    @kotlin.jvm.JvmDefault
    override fun asFloat(): Float {
        return pri()
    }

    companion object {

        fun histogram(pp: Iterable<Prioritized>, x: FloatArray): FloatArray {
            val bins = x.size
            val total = doubleArrayOf(0.toDouble())

            for (y in pp) {
                if (y == null)
                    continue
                var p = y.priElseZero()
                if (p > 1f) p = 1f
                val b = Util.bin(p, bins)
                x[b]++
                total[0]++
            }

            val t = total[0]
            if (t > 0.toDouble()) {
                for (i in 0 until bins)
                    x[i] = (x[i].toDouble() / t).toFloat()
            }
            return x
        }

        /**
         * double[histogramID][bin]
         */
        fun <X, Y> histogram(pp: Iterable<PriReference<Y>>, each: BiConsumer<PriReference<Y>, Array<DoubleArray>>, d: Array<DoubleArray>): Array<DoubleArray> {

            for (y in pp) {
                each.accept(y, d)
            }

            for (e in d) {
                var total = 0.0
                for (v in e) {
                    total += v
                }
                if (total > 0.toDouble()) {
                    var i = 0
                    val eLength = e.size
                    while (i < eLength) {
                        val f = e[i]
                        e[i] /= total
                        i++
                    }
                }
            }

            return d
        }
    }
}

